Merge mip 69

rdenarie opened the pull request:
<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->


<!-- Tips : 
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

hakermi committed:
feat: Implement Argon2PasswordEncoder - EXO-59676  - Meeds-io/MIPs#69 (#633)

Prior to this change, current password hashing algorithm is SHA-256 with
a salt of 64 bits generated by a random algorithm (SHA1PRNG). Our need
to a stronger hashing mechanism that will complexify and slow attacks
has lead us to Argon2 which is more pertinent for a password hashing
algorithm as recommended by ANSSI.
In This PR we implement a new password encoder that uses argon2 to
encode and hash the passwords

hakermi committed:
feat: Update the default password hashing algorithm - EXO-65150  - Meeds-io/MIPs#69 (#634)

Prior to this change, Old created passwords are hashed by the new
encoder algorithm and need un upgrade mechanism to re-hash theses
password without asking users to create new ones.
This PR is a part of the upgrade mechanism, will check the password slat
attribute and validate it by the proper algorithm, if it's hashed by the
old algorithm then it will be updated by the new one.

rdenarie committed:
feat: Update the default password hashing algorithm - EXO-65150  - Meeds-io/MIPs#69 (#636)

Before this change, if the old algo is still configured, users are not
able to login In addition, if it is changed to the new default algo
Argon2id, the user is still not able to connect In this change we do 2
things :
- continue to use the old algo if it is still configured
- make login working if the old algo was HashingEncoder/MD5, and if the
new algo is used.

<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new
feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses
questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->


<!-- Tips : 
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other
resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

hakermi committed:
feat: Update password validation in authentication phase regarding the new hash algorithm upgrade - EXO-65151 - Meeds-io/MIPs#69 (#639)

Prior to this change, As a final step in password hash upgrade old
authentication process need to be updated to take into consideration the
modification in passwords hashing.
This PR updates the authentication process to validate passwords by the
new hash algorithm including the old hashed passwords after an UP that
has updated the old hashing

rdenarie committed:
fix: Improve security for rememberme feature - EXO-59675 - meeds-io/MIPs#69

Before this fix, user password was encoded in the rememberme token, with
a reversible function This commit update the rememberme to not use the
user password in the rememberme token

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/PicketLinkIDMServiceImpl.java:
```diff
@@ -210,6 +212,16 @@ public IdentitySession getIdentitySession(String realm) throws Exception {
+
+    @Override
+    public ExtendedAttributeManager getExtendedAttributeManager() throws Exception {
+      return new ExtendedAttributeManager((IdentitySessionImpl) getIdentitySession());
```
Shouldn't we avoid building anew object each time this getter is invoked ?

rdenarie replied:
Done.

boubaker commented on component/identity/src/main/java/org/gatein/portal/idm/impl/store/attribute/ExtendedAttributeManager.java:
```diff
@@ -0,0 +1,131 @@
+
+  private static PicketLinkIDMService idmService;
+
+  public ExtendedAttributeManager(IdentitySessionImpl session) {
```
Can we make this as a service and move `session` parameter into `validatePassword` method as parameter ?

rdenarie replied:
ExtentedAttributeManager extends AttributeManagerImpl, which need IdentitySessionImpl in constructor. So ExtendedAttributeManager have to use a constructor with IdentitySessionImpl, to call super(session).

For this reason, I think that it is not possible to change it to a service. 
As you advise, I change in PicketLinkIDMService by reusing the same object, instead of instanciate one each time.

boubaker replied:
Thanks

boubaker commented on component/identity/src/test/resources/conf/picketlink/exo.portal.component.identity-picketlink-idm-config.xml:
```diff
@@ -30,6 +30,24 @@
       <identity-type-mappings>
         <user-mapping>USER</user-mapping>
       </identity-type-mappings>
+      <options>
+        <option>
+          <name>template</name>
+          <value>true</value>
+        </option>
+        <option>
+          <name>credentialEncoder.class</name>
+          <value>org.picketlink.idm.impl.credential.DatabaseReadingSaltEncoder</value>
+        </option>
+        <option>
+          <name>credentialEncoder.hashAlgorithm</name>
+          <value>SHA-256</value>
+        </option>
+        <option>
+          <name>credentialEncoder.secureRandomAlgorithm</name>
+          <value>SHA1PRNG</value>
+        </option>
+      </options>
```
Should move this into `supported-identity-object-types`/`<identity-object-type>`/`<options>` of `HibernateStore`.
In fact, when declaring `HibernateStore` and other stores, like `LDAP`, this option will have to be defined for Hibernate only. BTW, please update other PLIDM config files in the same folder, section `HibernateStore`.

rdenarie replied:
In picketlink, the encoder is an option of the realm, not of the store.

I understand the concern, and share, but PL seems not allowing it

boubaker replied:
> In picketlink, the encoder is an option of the realm, not of the store.

ok I see. In PLIDM Ldap Store source code, it seems not used. For my information, did you tested on LDAP conf ?

rdenarie replied:
I didnt test as LDAP connexion do not hash password

rdenarie replied:
I just test, and the login is ok

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/RememberMeLoginModule.java:
```diff
@@ -0,0 +1,88 @@
+
+      return true;
+    } catch (final Exception e) {
+      log.warn(e.getMessage());
+      throw new LoginException(e.getMessage());
```
either log or throw exception, never both

rdenarie replied:
Done.

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/RememberMeLoginModule.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Override
+  public boolean login() throws LoginException {
+    log.debug("In login of RemembermeLoginModule.");
```
useless ?

```suggestion
```

rdenarie replied:
Done

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/RememberMeLoginModule.java:
```diff
@@ -0,0 +1,88 @@
+      }
+      String token = LoginUtils.getRememberMeTokenCookie(servletRequest);
+
+      if (token!=null) {
```
formatting (please format the whole file)

```suggestion
      if (token != null) {
```

boubaker replied:
please format the whole file

rdenarie replied:
Done

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/RememberMeLoginModule.java:
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2013 eXo Platform SAS.
```
Wrong License Header
```suggestion
```

rdenarie replied:
Done

boubaker commented on component/identity/src/main/java/org/gatein/portal/idm/impl/store/attribute/ExtendedAttributeManager.java:
```diff
@@ -0,0 +1,131 @@
+
+  @Override
+  public boolean validatePassword(User user, String password) throws IdentityException {
+    Attribute salt128 = getAttribute(user.getKey(), PASSWORD_SALT_USER_ATTRIBUTE);
```
We don't need to add the length of salt in variable names for better future code upgradability

```suggestion
    Attribute salt = getAttribute(user.getKey(), PASSWORD_SALT_USER_ATTRIBUTE);
```

rdenarie replied:
Done

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/security/security/CookieTokenService.java:
```diff
@@ -135,34 +77,29 @@ public void start() {
             try {
-                this.tokenStore.saveToken(new GateInTokenStore.TokenData(id, hashedRandomString, encodedCredentials, new Date(expirationTimeMillis), type));
-//                tokenContainer.saveToken(context.getSession(), id, hashedRandomString, encodedCredentials, new Date(expirationTimeMillis, tokenType));
+                this.tokenStore.saveToken(new GateInTokenStore.TokenData(selector, hashedRandomString,
```
A security layer has been removed here when removing the usage of `CodecInitializer` to encrypt part of the content. In fact, we 'll need to preserve the usage of `CodecInitializer` to encrypt the stored Token in database.
In this case, when someone accesses the database, he still can't access users accounts until he get access to the `codecKey.txt` file from server.

The `hashedRandomString` has to be encrypted as well IMO and never add it in cookie or expose it as plain text. (it's similar to the password itself).

boubaker commented on webui/portal/src/main/java/org/exoplatform/portal/webui/login/UIForgetPassword.java:
```diff
@@ -126,8 +127,7 @@ public void execute(Event<UIForgetPassword> event) throws Exception {
             RemindPasswordTokenService tokenService = uiForm.getApplicationComponent(RemindPasswordTokenService.class);
-            Credentials credentials = new Credentials(user.getUserName(), "");
-            tokenId = tokenService.createToken(credentials,tokenService.FORGOT_PASSWORD_TOKEN);
+            tokenId = tokenService.createToken(user.getUserName(), CookieTokenService.FORGOT_PASSWORD_TOKEN);
```
This class isn't used anymore, shouldn't we delete it ?

rdenarie replied:
Do you mean in import ? 
If yes, I remove it.

boubaker replied:
In fact, the whole WebUI class should be removed since it's not used in a WebUI Portlet, isn't it ?

rdenarie replied:
I removed the complete 'login' folder

boubaker commented on component/web/security/pom.xml:
```diff
@@ -61,6 +61,10 @@
       <groupId>org.gatein.captcha</groupId>
       <artifactId>simplecaptcha</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcprov-jdk15on</artifactId>
+    </dependency>
```
Can you please move this into module `exo.portal.component.api` since it's a generic dependency used by multiple modules.

```suggestion
```

rdenarie replied:
Done

hakermi committed:
fix: Add passwordSalt128 attribute to non profile keys - EXO-65823 - Meeds-io/meeds#1061 (#643)

After adding new passwordSalt128 attribute for the new hashing
algorithm, the attribute was returned in gatein-profile and added inside
social profile.
This PR adds the attribute to the non profile attributes list, to
prevent it from being returned in gatein-profile.

rdenarie committed:
fix: Improve security for rememberme feature - EXO-59675 - meeds-io/MIPs#69

This commit fix feedback on the merge

rdenarie commented:
@boubaker 
Can you rewiew one more time ? 
I add a modification to remove unused import and to fix sonar feedback. 
Thanks
